Change the way of printing logs:

CedricHerzog opened the pull request:
The API calls were not printed in the right order.
This commit introduces a string that we fill for each API call so we can return all the information at once.

Outputing arguments in wide strings format still needs to be implemented.
For now, only the address is printed.

---
As you can see in the picture bellow, I had problems in the way the tool prints logs.
I wrote something to correct it and it seems to work for the API calls' names.
As I do not need the information about the arguments, I did not test whether or not they are well printed.  
I'm not used to code in C++ so this commit may not follow the good practices. I thought you might be interested anyway.

![error](https://user-images.githubusercontent.com/6296846/89924720-b7ce0900-dc02-11ea-8b14-3da373ce32e2.PNG)

Commit:
Change the way of printing logs:
The API calls were not printed in the right order.
This commit introduces a string that we fill for each API call so we can return all the information in once.

Outputing arguments in wide strings format still needs to be implemented.
For now, only the address is printed.

mxmssh commented:
Thank you for your commit! :) I believe it is a usefull feature but I see that it is not passing build tests. We first need to make sure that the pull request is not breaking anything to move forward.

mxmssh commented on drltrace_src/drltrace.cpp:
```diff
@@ -2,47 +2,47 @@
- * The runtime options for this client are specified in drltrace_options.h,
- * see DROPTION_SCOPE_CLIENT options.
- */
+ /*
```
Additionally, I believe your IDE auto-formated the file which makes code review very hard. drltrace follows DynamoRIO's code standard. We would prefer to keep it as is.

CedricHerzog replied:
Yes you're right i'll remove the changes made by the auto-format in a new commit.

mxmssh commented on drltrace_src/drltrace.cpp:
```diff
@@ -80,85 +80,114 @@ static std::vector<std::string> filter_module_blacklist;
 {
     bool pointer = !TEST(DRSYS_PARAM_INLINED, arg->mode);
-    dr_fprintf(outf, pointer ? PFX : (leading_zeroes ? PFX : PIFX), arg->value);
+    char str[128];
```
What if string is longer than 128 bytes?

mxmssh commented on drltrace_src/drltrace.cpp:
```diff
@@ -80,85 +80,114 @@ static std::vector<std::string> filter_module_blacklist;
-              (arg->type_name == NULL ||
-              TESTANY(DRSYS_PARAM_INLINED|DRSYS_PARAM_RETVAL, arg->mode)) ? "" : "*",
-              arg->size);
+    char str[128];
```
same here

mxmssh replied:
and in a lot of places below.

CedricHerzog replied:
As I fill it with a string representing a pointer address, I thought that 128 should be more than enough to contain it.
To fill it I could use snprintf instead of sprintf to avoid overflow?

mxmssh commented on drltrace_src/drltrace.cpp:
```diff
@@ -300,78 +335,94 @@ lib_entry(void *wrapcxt, INOUT void **user_data)
 
-	/* If the blacklist entry contains a wildcard, then compare only the shortest
-	 * part of either string. */
+        /* If the whitelist entry contains a wildcard, then compare only the shortest
```
Are you sure you are using the latest code for this commit? I see a bunch of code removed for some reason...

CedricHerzog replied:
Yes, my master was up to date with your master.
It may be just due to the re-format.

Commit:
- Remove changes of the previous commit due to auto-format
- Use snprintf instead of sprintf
- PFX seems to return the address in capital letters when using snprintf on Windows
- PFX also removes 0x at the beginning when using snprintf on Windows

Commit:
Correct error of compilation on linux

Commit:
- Try to correct the formatting of hex addresses
- Changes a sprintf to a snprintf I've forgotten
- Correct the outputing of string values %s and %S

Commit:
- Change the removal of the last char in the removal of \n and \r if exist

Commit:
Minor corrections to output logs using %s

CedricHerzog commented:
Hi,

I can't manage to make my code pass the 4th windows test.
Do you have an idea why it fails?

mxmssh commented:
I think the problem is this difference between correct and incorrect log printing:
Before it was
```
4: ~~1412~~ MSVCR120.dll!printf
4:     arg 0: done
4:  (type=char*, size=0x0)
```
after
```
4: ~~2952~~ MSVCR120.dll!printf
4:     arg 0: done (type=char*, size=0x0)
```
You can just change expected output regex in the test to reflect the new way of printing. It is here

https://github.com/mxmssh/drltrace/blob/94dc91107dcf10a99a715947494f38c6de11e6b0/drltrace_src/CMakeLists.txt#L465

You need to remove ```\n``` after done.

Commit:
Keeps the \r\n in arguments to pass tests

CedricHerzog commented:
> 
> 
> I think the problem is this difference between correct and incorrect log printing:
> Before it was
> 
> ```
> 4: ~~1412~~ MSVCR120.dll!printf
> 4:     arg 0: done
> 4:  (type=char*, size=0x0)
> ```
> 
> after
> 
> ```
> 4: ~~2952~~ MSVCR120.dll!printf
> 4:     arg 0: done (type=char*, size=0x0)
> ```
> 
> You can just change expected output regex in the test to reflect the new way of printing. It is here
> 
> https://github.com/mxmssh/drltrace/blob/94dc91107dcf10a99a715947494f38c6de11e6b0/drltrace_src/CMakeLists.txt#L465
> 
> You need to remove `\n` after done.

Ok that was indeed the problem.
I changed the code to stick to the tests.

I still have to do some modifications as I have warning during the build.
Then, it should be good to be reviewed.

mxmssh commented:
Ok, ping me whenever you are ready for review with this :)

CedricHerzog commented:
The warnings are due to the usage of %x instead of %p to print pointers.
It may behave differently on some systems. Do you know how to fix this warnings properly and keep the same behavior?

I tried to use the PFX and PIFX but it does not output the same string with snprintf() than with dr_fprintf().
